### PR TITLE
fix(mcp): validate server names to prevent path traversal

### LINKED ISSUE #352: mcp: server name used in temp dir path without validation allows path traversal
## Summary

Found as an OOS finding during review of the `feat/mcp-sandbox` branch.

## Location

`crates/astrid-mcp/src/server.rs` line 374:

```rust
.unwrap_or_else(|| std::env::temp_dir().join("astrid-mcp").join(name));
```

## Impact

The server `name` field is taken directly from the TOML hash key in `servers.toml` (e.g., `[servers.my-server]`) and used to construct a file system path without any validation. A crafted name like `../../etc` would resolve `std::env::temp_dir().join("astrid-mcp").join("../../etc")` to a path **outside** the intended temp directory (e.g., `/tmp/etc` or `/private/var/etc` on macOS).

`create_dir_all` is then called on this path, which would create arbitrary directories. On Linux, this path is also passed to `bwrap --bind`, allowing the sandboxed process write access to a directory outside the intended writable root.

The same issue applies to the dynamic `add_server` path where `server_id` comes from `format\!("capsule:{}", self.manifest.package.name)` (capsule manifest data).

## Severity

Important - requires crafted `servers.toml` or malicious capsule manifest, but creates directories at attacker-controlled paths and grants sandbox write access to those paths.

## Suggested Fix

Validate the server name before using it in path construction:

```rust
fn validate_server_name(name: &str) -> McpResult<()> {
    if name.contains('/') || name.contains('\\') || name.contains('.') && name.starts_with('.') {
        return Err(McpError::ConfigError(format\!(
            "Server name must not contain path separators or start with '.': {name}"
        )));
    }
    Ok(())
}
```

Or use a hash/sanitized form of the name for the directory path.

### PR DESCRIPTION
**Pull request by joshuajbouw**
## Summary

- Add `validate_server_name()` that enforces an ASCII allowlist (alphanumeric, hyphens, underscores, colons, dots) with leading-dot rejection, 128-char max length, and empty-string check
- Enforce validation at all entry points: `ServersConfig::load()`, `ServersConfig::add()`, `ServerManager::start()`, and `ServerManager::add_server()`
- Add 10 unit tests covering valid names, path traversal, leading dots, null bytes, Unicode lookalikes, shell metacharacters, length limits, and TOML config rejection

## Test Plan

- `cargo test -p astrid-mcp` - all 96 unit tests + 2 doctests pass
- `cargo test --workspace` - full workspace green
- `cargo clippy --workspace -- -D warnings` - clean
- New tests exercise every risk from the threat model:

| Risk | Test function |
|---|---|
| Path traversal (`../`) | `validate_server_name_rejects_path_traversal` |
| Hidden dirs (leading `.`) | `validate_server_name_rejects_leading_dot` |
| Null byte injection | `validate_server_name_rejects_null_byte` |
| Length overflow | `validate_server_name_rejects_too_long` |
| Empty name | `validate_server_name_rejects_empty` |
| Unicode lookalikes | `validate_server_name_rejects_unicode_lookalikes` |
| Shell metacharacters | `validate_server_name_rejects_special_chars` |
| TOML config entry point | `load_rejects_traversal_name_in_toml` |
| `add()` entry point | `add_rejects_traversal_name` |
| Valid names accepted | `validate_server_name_accepts_valid_names` |

## Breaking Changes

`ServersConfig::add()` now returns `McpResult<()>` instead of `()`. All internal callers updated.

## Related Issues

Closes #352

### CODE REVIEW

File: crates/astrid-mcp/src/config.rs [hunk 1]
```diff
@@ -212,6 +212,49 @@ impl ServerConfig {
...
+        )));
+    }
+    if name.starts_with('.') {
+        return Err(McpError::ConfigError(format!(
+            "server name must not start with '.': {name}"
+        )));
+    }
+    if !name
+        .chars()
+        .all(|c| c.is_ascii_alphanumeric() || matches!(c, '-' | '_' | ':' | '.'))
```
[the remark this answers is not in the document]
Comment (joshuajbouw) [reply crates/astrid-mcp/src/config.rs]: @gemini-code-assist Good catch - applied. Switched to `.bytes().all(|b| b.is_ascii_alphanumeric() || matches\!(b, b'-' | b'_' | b':' | b'.'))` to avoid unnecessary UTF-8 decoding since the entire allowlist is ASCII.

File: crates/astrid-mcp/src/config.rs [hunk 2]
```diff
@@ -212,6 +212,57 @@ impl ServerConfig {
...
+/// Returns [`McpError::ConfigError`] if the name is empty, too long, starts with
+/// a dot, or contains characters outside the allowed set.
+pub fn validate_server_name(name: &str) -> McpResult<()> {
+    // Truncate the displayed name in error messages to prevent log poisoning
+    // from attacker-controlled input.
+    let display_name = if name.len() > 40 {
+        format!("{}...", &name[..name.floor_char_boundary(40)])
+    } else {
+        name.to_string()
+    };
```
[the remark this answers is not in the document]
Comment (joshuajbouw) [reply crates/astrid-mcp/src/config.rs]: @gemini-code-assist Applied - switched to Cow::Borrowed(name) for the short-name path to avoid the allocation.

[the remark this answers is not in the document]
Comment (joshuajbouw) [reply crates/astrid-mcp/src/config.rs]: @gemini-code-assist Duplicate of the above - applied in the same commit.

File: crates/astrid-mcp/src/server.rs [hunk 3]
```diff
@@ -179,6 +181,8 @@ impl ServerManager {
     /// # Errors
     /// Returns an error if the server is already running.
     pub async fn add_server(&self, name: &str, config: ServerConfig) -> McpResult<()> {
+        crate::config::validate_server_name(name)?;
```
[the remark this answers is not in the document]
Comment (joshuajbouw) [reply crates/astrid-mcp/src/server.rs:235]: @gemini-code-assist Good catch. Now validating both `name` and `config.name` in `add_server()` to prevent a mismatched valid-name / malicious-config.name attack vector.

File: crates/astrid-mcp/src/config.rs [hunk 4]
```diff
@@ -212,6 +212,57 @@ impl ServerConfig {
...
+///
+/// # Errors
+///
+/// Returns [`McpError::ConfigError`] if the name is empty, too long, starts with
+/// a dot, or contains characters outside the allowed set.
+pub fn validate_server_name(name: &str) -> McpResult<()> {
+    // Truncate the displayed name in error messages to prevent log poisoning
+    // from attacker-controlled input.
+    let display_name: std::borrow::Cow<'_, str> = if name.len() > 40 {
+        std::borrow::Cow::Owned(format!("{}...", &name[..name.floor_char_boundary(40)]))
```
[the remark this answers is not in the document]
Comment (joshuajbouw) [reply crates/astrid-mcp/src/config.rs]: @gemini-code-assist Applied. Extracted the magic number 40 into a `MAX_NAME_DISPLAY_LEN` constant at module level.

File: crates/astrid-mcp/src/server.rs [hunk 5]
```diff
@@ -230,6 +232,9 @@ impl ServerManager {
...
+        crate::config::validate_server_name(&config.name)?;
```
[the remark this answers is not in the document]
Comment (joshuajbouw) [reply crates/astrid-mcp/src/server.rs]: @gemini-code-assist Applied. Replaced the dual-validate with a single validate + equality check. If name and config.name differ, we now fail with a clear mismatch error instead of silently storing inconsistent identity.
